### PR TITLE
Hyphenation patterns update (cleaned up #376)

### PR DESCRIPTION
**Pull request by poire-z**
See #376 for the whole discussion and work steps.
This is just #376 with the 33 work commits squashed into 17 logical commits.

---
This change is [Reviewable](https://reviewable.io/reviews/koreader/crengine/378)

### CODE REVIEW

**Comment by poire-z** [comment]
Pinging @roshavagarga and @cramoisi (I can't add you as reviewers...) for a quick look and confirmation that I didn't mess anything, before merging.

**Comment by cramoisi** [comment]
I will look into it after work :)

File: cr3gui/data/hyph/languages.json [hunk 1]
```diff
@@ -149,6 +149,14 @@
         "language": "it",
         "aliases": ["ita"]
     },
+    {
+        "name": "Latvian",
+        "filename": "Latvian.pattern",
+        "language": "lv",
+        "left_hyphen_min": 2,
+        "right_hyphen_min": 2
```
Comment (roshavagarga) [inline cr3gui/data/hyph/languages.json:167]: I missed adding a comma after this line (157)

Comment (poire-z) [reply cr3gui/data/hyph/languages.json:167]: Yup, just saw that, going to fix it.

**Comment by roshavagarga** [comment]
I did my once-over and it seems fine to me, though I did notice this has 1 less new line compared to the original, but no idea where that 1 missing line is from, probably something I missed and you fixed?

**Comment by poire-z** [comment]
Yes, a last trailing blank line in Esperanto.pattern that I removed (to avoid github highlighting it as "not clean").

**Comment by roshavagarga** [comment]
Ah, those always annoyed me, good to know I can get rid of them :+1:

**Comment by poire-z** [comment]
Added 2 commits to rename badly named Romanian and Ukrainian pattern files, and some methods to help cleaning up frontend readertypography.lua code.

**Comment by roshavagarga** [comment]
@poire-z Not sure how to add changes to this PR, so I've attached the latest Spanish pattern :)

[Spanish.zip](https://github.com/koreader/crengine/files/5238257/Spanish.zip)

**Comment by poire-z** [comment]
OK, added.

> I tried learning git some time ago and just gave up out of frustration and stuck with using github's web ui.

If you plan to do more work/fixes like these, and you have a linux machine / can do git command line stuff, I can give you a small cheatsheet on how to proceed to make simple PRs on new branches (that I could more easily reword, but even you, and we'll be here to help if you get stuck).

**Comment by roshavagarga** [comment]
Yeah, the one Linux machine I have is a decrepit single-core tiny notebook from like 7-8 years ago that's running Lubuntu, because nothing else runs well on it :/
I generally stick to Windows 7, because gaming, but you can still share the cheatsheet and if I do resort to it I can always boot from a usb drive or something.

**Review by cramoisi — CHANGES_REQUESTED** [review]

File: cr3gui/data/hyph/Spanish.pattern [hunk 2]
```diff
@@ -3239,43 +4038,108 @@
...
   <pattern> reu3mati</pattern>
   <pattern> reu3máti</pattern>
-  <pattern> re3unir</pattern>
-  <pattern> re3unír</pattern>
+  <pattern> reu3nir</pattern>
+  <pattern> reu3nír</pattern>
+  <pattern> reu3nia</pattern>
+  <pattern> reu3ní</pattern>
+  <pattern> reu3nis</pattern>
+  <pattern> reunim</pattern>
```
[the remark this answers is not in the document]
Comment (roshavagarga) [reply cr3gui/data/hyph/Spanish.pattern:4098]: Only the ones without numbers @ 4098, right?

[the remark this answers is not in the document]
Comment (cramoisi) [reply cr3gui/data/hyph/Spanish.pattern:4098]: yes :)

**Comment by poire-z** [comment]
(@roshavagarga : I'll do the requested modifications - later this afternoon.)

**Comment by poire-z** [comment]
OK, some quick notes (build from my own notes, removing more complex and rare stuff):

```
FORK repo on github

CLONE AND SETUP UPSTREAM :
$ git clone https://github.com/yourname/crengine.git
$ git remote -v
$ git remote add upstream https://github.com/koreader/crengine.git
$ git remote -v
    origin  https://github.com/yourname/crengine.git (fetch)
    origin  https://github.com/yourname/crengine.git (push)
    upstream  https://github.com/koreader/crengine.git (fetch)
    upstream  https://github.com/koreader/crengine.git (push)

NEVER ADD STUFF TO YOUR master BRANCH, it will just be kept up to date
with upstream master as a reference.

To sync your local master branch with upstream:
$ git checkout master
$ git fetch upstream
$ git merge upstream/master
To update your remote branch with your now-sync'ed local branch:
$ git push origin master

----
CREATE A NEW BRANCH to be pushed as a PR:
$ git checkout -b my-mod-branch
<< edit files >>
To see your changed stuff
$ git status
$ git diff

To make a commit with your changed stuff:
$ git add path/to/file
$ git add .    (if you want to add them all)
$ git status
$ git commit
<< lauches vi >>
    Your commit title

    Your commit details if needed.
Check all looks good:
$ git log
$ git log -p
Add more commits if needed.

When ready, push for a PR
$ git push -u origin my-mod-branch         (-u to create it)
And follow the url to github to open a PR.

Add more commits (as above) if needed to answer review comments, and push them:
$ git push origin my-mod-branch

Here, you would have a lot of other options, to rebase, squash, remove commits,
I could detail them when needed.

When PR is merged, cleanup and resync everything:
$ git checkout master
$ git branch -D my-mod-branch
$ git fetch --prune
$ git fetch upstream ; git merge upstream/master
$ git push origin master

Ready for next PR :)
```

File: cr3gui/data/hyph/Spanish.pattern [hunk 3]
```diff
@@ -3239,43 +4038,108 @@
...
+  <pattern> re3áme</pattern>
+  <pattern> re3amé</pattern>
+  <pattern> re3amo</pattern>
+  <pattern> re3ámo</pattern>
+  <pattern> re3amó</pattern>
   <pattern> re3aparec</pattern>
+  <pattern> re3aprend</pattern>
+  <pattern> re3aprénd</pattern>
+  <pattern> re3apret</pattern>
+  <pattern> reapríet</pattern>
```
[the remark this answers is not in the document]
Comment (poire-z) [reply cr3gui/data/hyph/Spanish.pattern:4082]: All 4 ? Or just the last one without any number?

[the remark this answers is not in the document]
Comment (poire-z) [reply cr3gui/data/hyph/Spanish.pattern:4082]: I'll assume it's just the last one.

[the remark this answers is not in the document]
Comment (poire-z) [reply cr3gui/data/hyph/Spanish.pattern:4082]: Of, right, you answered that below :)

[the remark this answers is not in the document]
Comment (cramoisi) [reply cr3gui/data/hyph/Spanish.pattern:4082]: yup. only the non-numbered one. I messed with my line selection, sorry :)

**Comment by poire-z** [comment]
Ok, done. Please have a quick look again as I did this in a hurry.

**Review by cramoisi — APPROVED** [review]
ok good for me :)

**Review by roshavagarga — APPROVED** [review]

**Comment by virxkane** [comment]
This file is sufficient to describe the 'TeX Hyphenation patterns' format?
https://github.com/hyphenation/tex-hyphen/blob/master/docs/tb87nemeth.pdf

Ok, I can pick hyphenation patterns files from KOReader into CoolReader. Thanks to all contributors.
But I thought, what to do with the "``*_hyphen_(Alan) .pdb``" files, just delete it? But what if there is something missing in the "``*.pattern``" files?
So, I wrote a simple converter (dumper) for "``*_hyphen_(Alan).pdb``" or "``*.pattern``" files and saw that some of the patterns that are present in "``*_hyphen_(Alan).pdb``" are missing in "``*.pattern``".
I guess that "``*_hyphen_(Alan).pdb``" files transfered from AlReader program: https://github.com/koreader/crengine/blob/a00d1b8c26f4a621d805575a6a4cf10d81c8e7c7/crengine/src/hyphman.cpp#L2
And I don't know where Alan got these patterns and how correct they are.
But if I make dump from "``English_US_hyphen_(Alan).pdb``" I got 295799 bytes instead of 126141 from file ``English_US.pattern``. And this fact makes you wonder which file is better and more correct.
I found that in KOReader "``*_hyphen_(Alan).pdb``" files dropped in https://github.com/koreader/crengine/pull/209 and https://github.com/koreader/crengine/pull/206 and I guess nobody checked contents of the "``*_hyphen_(Alan).pdb``" files.
Maybe we can combine them somehow? What is the best way to do this? Is it ok to just add omitted patterns in "``*.pattern``" from "``*_hyphen_(Alan).pdb``"?

P.S. This tool can be found in my not-yet-merged branch here: https://github.com/virxkane/coolreader/tree/updates-2020.09-1/crengine/Tools/HyphDumper
I would be grateful if someone points out the errors.

**Comment by cramoisi** [comment]
I could check the differences if you provide a human readable dump of English pdf file. But as now we are following the hyphen patterns from recognised sources, so I don't think we need to worry about trying to make better than linguists already did for libreoffice, etc. I think feedback from native speakers should be enough as it's somehow tricky to edit these files without in dept knowledge of the languages...

**Comment by virxkane** [comment]
How can I make "human readable dump of English pdf file"? I can only convert ``*_hyphen_(Alan).pdb`` into ``*.pattern``.

**Comment by cramoisi** [comment]
That will do as I'm human and I can read a pattern file with my eyes :)

Btw : I checked the content of pdb files back then but only for a few of them. As I said, some differences are legit

**Comment by poire-z** [comment]
> Maybe we can combine them somehow? What is the best way to do this? Is it ok to just add omitted patterns in "*.pattern" from "*_hyphen_(Alan).pdb"?

I don't think combining them makes any sense: it's not really "the more items the better", but having a set that is consistent within itself (some patterns with numbers overriding some others with other numbers - mixing sources could possibly have these numbers relations incoherent).
Given that we don't really know who made the _Alan.pdb from what and when, and that the sources used to build the *.pattern are recent, referenced and updated, our *.pattern are a lot more trustable.

**Comment by virxkane** [comment]
> I don't think combining them makes any sense: it's not really "the more items the better", but having a set that is consistent within itself (some patterns with numbers overriding some others with other numbers - mixing sources could possibly have these numbers relations incoherent).
> Given that we don't really know who made the _Alan.pdb from what and when, and that the sources used to build the *.pattern are recent, referenced and updated, our *.pattern are a lot more trustable.

Yes, it looks logical.
And finally, for the connoisseurs of English, which I am not, dump 4 files:
1. https://github.com/koreader/crengine/blob/master/cr3gui/data/hyph/English_GB.pattern ->
[English_GB_dump.pattern.zip](https://github.com/koreader/crengine/files/5253472/English_GB_dump.pattern.zip)
2. https://github.com/koreader/crengine/blob/master/cr3gui/data/hyph/English_US.pattern -> 
[English_US_dump.pattern.zip](https://github.com/koreader/crengine/files/5253473/English_US_dump.pattern.zip)
3. https://github.com/buggins/coolreader/blob/master/cr3gui/data/hyph/English_GB_hyphen_(Alan).pdb -> 
[English_GB_Alan_dump.pattern.zip](https://github.com/koreader/crengine/files/5253474/English_GB_Alan_dump.pattern.zip)
4. https://github.com/buggins/coolreader/blob/master/cr3gui/data/hyph/English_US_hyphen_(Alan).pdb -> 
[English_US_Alan_dump.pattern.zip](https://github.com/koreader/crengine/files/5253475/English_US_Alan_dump.pattern.zip)

The dump for "``*.pattern``" is the same as original, but all patterns are sorted for ease of comparison.
* Difference between "English_GB_Alan_dump.pattern" and "English_US_dump.pattern" is less than between "English_US_Alan_dump.pattern" and "English_US_dump.pattern" :).
* Patterns in "English_GB_hyphen_(Alan).pdb" like "a8s7s8o7c8i8a8t8e8s" look like a hack.

Is there something useful about these files "``*_hyphen_(Alan).pdb``" ?

**Comment by cramoisi** [comment]
Ok, I checked and actual patterns are more developed and consistent than the old one. A good example is that exceptions introduced by the hacks you mentioned in the pdb files are all processed correctly in current pattern file. So I guess you could get rid of these _(alan).pdb files once for all :-)

**Comment by virxkane** [comment]
@cramoisi Thank you.

**Pull request merged**